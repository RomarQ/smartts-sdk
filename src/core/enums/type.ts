export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
