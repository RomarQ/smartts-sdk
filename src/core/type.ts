import { IToString } from '../typings';

export class BaseType {
    private _ = 'Type';
}

export class SimpleType extends BaseType {
    constructor(public name: string) {
        super();
    }

    toString() {
        return `"${this.name}"`;
    }
}

export class T_TList extends BaseType {
    constructor(public innerType: IToString) {
        super();
    }

    toString() {
        return `(list ${this.innerType})`;
    }
}

export const TList = (innerType: BaseType) => new T_TList(innerType);

export const TNat = new SimpleType('nat');
export const TInt = new SimpleType('int');
export const TMutez = new SimpleType('mutez');
export const TString = new SimpleType('string');
