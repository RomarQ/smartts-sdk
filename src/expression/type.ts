import type { IExpression } from '../typings/expression';
import type { IType } from '../typings/type';
import { Expression } from '../core/expression';
import ExpressionAtom from '../core/enums/expression';
import { LineInfo } from '../misc/utils';
import { proxy } from '../misc/proxy';

/**
 * Annotate an expression with type.
 *
 * ```typescript
 * AsType(GetVariable("some_variable"), TNat());
 * ```
 *
 * @param expression
 * @param type
 * @param {LineInfo} line Source code line information (Used in error messages)
 *
 * @category | Type Handling
 *
 * @returns {IExpression} An expression
 */
export const AsType = (expression: IExpression, type: IType, line = new LineInfo()): IExpression<any> =>
    proxy(new Expression(ExpressionAtom.type_annotation, expression, type, line), Expression.proxyHandler);
