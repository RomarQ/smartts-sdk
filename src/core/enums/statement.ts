export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
