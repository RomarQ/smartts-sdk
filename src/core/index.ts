import type { IToString, IToType } from '../typings';
import type { ILiteral } from '../typings/literal';
import type { IType } from '../typings/type';

import { Expression } from './expression';
import { Mutez } from './literal';
import Utils, { LineInfo } from '../misc/utils';
import { TUnit } from './type';

export class Flag {
    args: string[] = [];

    constructor(public flag: string, ...args: string[]) {
        this.args = args;
    }

    toString() {
        return `(${this.flag}${(' ' + this.args.join(' ')).trim()})`;
    }
}

interface EntryPointOptions {
    mock?: boolean;
    lazy?: boolean;
    lazyAndCodeless?: boolean;
}
export class EntryPoint {
    options = {
        mock: false,
        lazy: false,
        lazyAndCodeless: false,
    };
    inType?: IType;
    statements: IToString[] = [];

    constructor(public name: string, public line = new LineInfo()) {}

    config(options?: EntryPointOptions) {
        if (options?.mock) {
            this.options.mock = options.mock;
        }
        if (options?.lazy) {
            this.options.lazy = options.lazy;
        }
        if (options?.lazyAndCodeless) {
            this.options.lazyAndCodeless = options.lazyAndCodeless;
        }
        return this;
    }

    inputType(type: IType) {
        this.inType = type;
        return this;
    }

    code(callback: (arg: IToString) => IToString[]) {
        this.statements = callback(new Expression('params', new LineInfo()));
        return this;
    }

    toString() {
        const notMock = Utils.capitalizeBoolean(!this.options.mock);
        const isLazy = Utils.capitalizeBoolean(this.options.lazy);
        const isLazyAndCodeless = Utils.capitalizeBoolean(this.options.lazyAndCodeless);
        const hasParams = Utils.capitalizeBoolean(!!this.inType);
        return `(${this.name} ${notMock} ${isLazy} ${isLazyAndCodeless} ${hasParams} ${
            this.line
        } (${this.statements.join(' ')}))`;
    }
}

export class Contract {
    options: {
        initialBalance: ILiteral;
        flags: Flag[];
    } = {
        initialBalance: Mutez(0),
        flags: [],
    };
    public initialStorage: IToString & IToType;
    public entries: EntryPoint[];

    constructor(
        args: {
            initialStorage: IToString & IToType;
            entries: EntryPoint[];
        },
        public line = new LineInfo(),
    ) {
        this.initialStorage = args.initialStorage;
        this.entries = args.entries;
    }

    config(options?: { initialBalance?: ILiteral; flags?: Flag[] }) {
        if (options?.flags) {
            this.options.flags = options.flags;
        }
        if (options?.initialBalance) {
            this.options.initialBalance = options.initialBalance;
        }
        return this;
    }

    toString() {
        return `
        (
            template_id (static_id 0 ${this.line})
            storage ${this.initialStorage.toString()}
            storage_type (${this.initialStorage.toType()})
            messages (${this.entries.map((entry) => entry.toString()).join(' ')})
            flags (${this.options.flags.map((flag) => flag.toString())})
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance ${this.options.initialBalance}
        )
        `;
    }
}

export { default as Command } from './command';
export { default as Expression } from './expression';
export { default as Type } from './type';
export { default as Literal } from './literal';
export * from './blockchain_operations';
