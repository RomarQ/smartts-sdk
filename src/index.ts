// Load polyfills needed by SmartML
import './compiler/polyfills';

export * from './core';
export * from './type';
export * from './expression';
export * from './statement';
