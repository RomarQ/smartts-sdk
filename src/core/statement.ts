import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { LineInfo } from '../misc/utils';
import StatementAtom from './enums/statement';
import { Expression } from './expression';
import ExpressionAtom from './enums/expression';
import { Proxied, proxy } from '../misc/proxy';
import { Add, GetVariable, Iterator, LessThanOrEqual } from '../expression';
import { NewVariable, SetValue } from '../statement';

export class Statement implements IStatement {
    args;

    constructor(public name: string, ...args: (IExpression | IType | LineInfo | string | ILiteral)[]) {
        this.args = args || [];
    }

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class IfStatment implements IStatement {
    #condition: IExpression;
    #thenStatements: IStatement[];
    #elseStatements: IStatement[];
    #line: LineInfo;

    constructor(
        condition: IExpression,
        line = new LineInfo(),
        thenStatements?: IStatement[],
        elseStatements?: IStatement[],
    ) {
        this.#condition = condition;
        this.#thenStatements = thenStatements || [];
        this.#elseStatements = elseStatements || [];
        this.#line = line;
    }

    public Then(statements: IStatement[]) {
        this.#thenStatements = statements;
        return this;
    }

    public Else(statements: IStatement[]) {
        this.#elseStatements = statements;
        return this;
    }

    private ifBlock() {
        return `(${StatementAtom.ifBlock} ${this.#condition} (${this.#thenStatements.join(' ')}) ${this.#line})`;
    }

    private elseBlock() {
        return `(${StatementAtom.elseBlock} (${this.#elseStatements.join(' ')}))`;
    }

    [Symbol.toPrimitive]() {
        let expression = this.ifBlock();
        if (this.#elseStatements.length) {
            expression = `${expression} ${this.elseBlock()}`;
        }
        return expression;
    }
}

export class VariantMatchStatement implements IStatement {
    static idCounter = 0;
    private cases: Record<string, IStatement[]> = {};

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(private variant: IExpression, private argumentName: string, private line = new LineInfo()) {}

    public setArgumentName(argumentName: string): this {
        this.argumentName = argumentName;
        return this;
    }

    public Case(branch: string, buildStatements: (arg: Proxied<IExpression>) => IStatement[]): this {
        const variantArgument = this.variantArgument(`${this.argumentName}_${branch}`, this.line);
        this.cases[branch] = buildStatements(variantArgument);
        return this;
    }

    private caseArgument(argumentName: string, line: LineInfo) {
        return new Expression(ExpressionAtom.cases_arg, `"${argumentName}"`, line);
    }

    private variantArgument(argumentName: string, line: LineInfo) {
        return proxy(new Expression(ExpressionAtom.variant_arg, `"${argumentName}"`, line), Expression.proxyHandler);
    }

    [Symbol.toPrimitive]() {
        const caseArgument = this.caseArgument(this.argumentName, this.line);
        const matchCases = Object.keys(this.cases).map((branch) => {
            return new Statement(
                StatementAtom.match,
                caseArgument,
                branch,
                `"${this.argumentName}_${branch}"`,
                `(${this.cases[branch].join(' ')})`,
                this.line,
            );
        });
        return `(${StatementAtom.match_cases} ${this.variant} "${this.argumentName}" (${matchCases.join(' ')}) ${
            this.line
        })`;
    }
}

export class WhileStatement implements IStatement {
    constructor(private condition: IExpression, private statements: IStatement[] = [], private line = new LineInfo()) {}

    public Do(statements: IStatement[]) {
        this.statements = statements;
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.whileBlock} ${this.condition} (${this.statements.join(' ')}) ${this.line})`;
    }
}

export class ForStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private from: IExpression,
        private to: IExpression,
        private increment: IExpression,
        private statements: IStatement[] = [],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = GetVariable(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        const variable = NewVariable(this.iteratorName, this.from);
        const condition = LessThanOrEqual(GetVariable(this.iteratorName), this.to);
        const stmts = [
            ...this.statements,
            SetValue(GetVariable(this.iteratorName), Add(GetVariable(this.iteratorName), this.increment)),
        ];
        return `${variable} (${StatementAtom.whileBlock} ${condition} (${stmts.join(' ')}) ${this.line})`;
    }
}

export class ForEachStatement implements IStatement {
    static idCounter = 0;

    static get nextID() {
        return ++VariantMatchStatement.idCounter;
    }

    constructor(
        private list: IExpression,
        private statements: IStatement[],
        private iteratorName: string,
        private line = new LineInfo(),
    ) {}

    public setIteratorName(iteratorName: string): this {
        this.iteratorName = iteratorName;
        return this;
    }

    public Do(callback: (iterator: Proxied<IExpression>) => IStatement[], line = new LineInfo()) {
        const iterator = Iterator(this.iteratorName, line);
        this.statements = callback(iterator);
        return this;
    }

    [Symbol.toPrimitive]() {
        return `(${StatementAtom.forGroup} "${this.iteratorName}" ${this.list} (${this.statements.join(' ')}) ${
            this.line
        })`;
    }
}
