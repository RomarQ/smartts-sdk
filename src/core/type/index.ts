import { composeRightCombLayout, parenthesis } from '../../misc/utils';
import { ILayout } from '../../typings/literal';
import { IType } from '../../typings/type';
import { Layout } from '../enums/layout';
import TypeAtom from '../enums/type';

export class SimpleType implements IType {
    _isType = true as const;

    constructor(public name: string) {}

    toString() {
        return `"${this.name}"`;
    }
}

export class ContainerType implements IType {
    _isType = true as const;

    constructor(public type: string, public innerTypes: (IType | number)[]) {}

    toString() {
        return `(${this.type} ${this.innerTypes.join(' ')})`;
    }
}

export class Type_Record implements IType {
    _isType = true as const;

    private layout: ILayout | Layout;

    constructor(private atom: string, private fields: Record<string, IType>, layout?: ILayout | Layout) {
        this.layout = layout || composeRightCombLayout(Object.keys(fields));
    }

    private translateLayout = (layout: ILayout | Layout): string => {
        const normalizeTuple = (layout: string | ILayout): string => {
            if (typeof layout === 'string') {
                return `("${layout}")`;
            }
            const [left, right] = layout;
            return parenthesis([normalizeTuple(left), normalizeTuple(right)].join(' '));
        };

        if (Array.isArray(layout)) {
            return layout.length ? `(Some ${normalizeTuple(layout)})` : 'None';
        }
        return '(Some Right)';
    };

    private buildFields = (fields: Record<string, IType>) => {
        return Object.entries(fields).map(([field, ty]) => `(${field} ${ty.toString()})`);
    };

    toString() {
        return `(${this.atom} (${this.buildFields(this.fields).join(' ')}) ${this.translateLayout(this.layout)})`;
    }
}

export const TUnknown: IType = {
    _isType: true as const,
    toString: () => '(unknown 0)',
};
// Singleton types
export const TUnit = () => new SimpleType(TypeAtom.unit);
export const TNat = () => new SimpleType(TypeAtom.nat);
export const TInt = () => new SimpleType(TypeAtom.int);
export const TMutez = () => new SimpleType(TypeAtom.mutez);
export const TString = () => new SimpleType(TypeAtom.string);
export const TBool = () => new SimpleType(TypeAtom.bool);
export const TAddress = () => new SimpleType(TypeAtom.address);
export const TTimestamp = () => new SimpleType(TypeAtom.timestamp);
export const TChain_id = () => new SimpleType(TypeAtom.chain_id);
export const TBytes = () => new SimpleType(TypeAtom.bytes);
export const TBls12_381_fr = () => new SimpleType(TypeAtom.bls12_381_fr);
export const TBls12_381_g1 = () => new SimpleType(TypeAtom.bls12_381_g1);
export const TBls12_381_g2 = () => new SimpleType(TypeAtom.bls12_381_g2);
export const TKey = () => new SimpleType(TypeAtom.key);
export const TKey_hash = () => new SimpleType(TypeAtom.key_hash);
export const TSignature = () => new SimpleType(TypeAtom.signature);
export const TOperation = () => new SimpleType(TypeAtom.operation);
export const TNever = () => new SimpleType(TypeAtom.never);
// Container types
export const TList = (innerType: IType) => new ContainerType(TypeAtom.list, [innerType]);
export const TSet = (innerType: IType) => new ContainerType(TypeAtom.set, [innerType]);
export const TOption = (innerType: IType) => new ContainerType(TypeAtom.option, [innerType]);
export const TMap = (keyType: IType, valueType: IType) => new ContainerType(TypeAtom.map, [keyType, valueType]);
export const TBig_map = (keyType: IType, valueType: IType) => new ContainerType(TypeAtom.big_map, [keyType, valueType]);
export const TPair = (left: IType, right: IType) => new ContainerType(TypeAtom.tuple, [left, right]);
export const TLambda = (left: IType, right: IType) => new ContainerType(TypeAtom.lambda, [left, right]);
export const TTicket = (innerType: IType) => new ContainerType(TypeAtom.ticket, [innerType]);
export const TContract = (innerType: IType) => new ContainerType(TypeAtom.contract, [innerType]);
export const TSapling_state = (memo: number) => new ContainerType(TypeAtom.sapling_state, [memo]);
export const TSapling_transaction = (memo: number) => new ContainerType(TypeAtom.sapling_transaction, [memo]);
// Artificial Types
export const TRecord = (fields: Record<string, IType>, layout?: ILayout | Layout) =>
    new Type_Record(TypeAtom.record, fields, layout);
export const TVariant = (fields: Record<string, IType>, layout?: ILayout | Layout) =>
    new Type_Record(TypeAtom.variant, fields, layout);

const Types = {
    TUnknown,
    // Singleton types
    TUnit,
    TNat,
    TInt,
    TMutez,
    TString,
    TBool,
    TBytes,
    TAddress,
    TTimestamp,
    TChain_id,
    TBls12_381_fr,
    TBls12_381_g1,
    TBls12_381_g2,
    TKey,
    TKey_hash,
    TSignature,
    TOperation,
    TNever,
    // Container types
    TList,
    TSet,
    TOption,
    TPair,
    TMap,
    TBig_map,
    TLambda,
    TTicket,
    TContract,
    TSapling_state,
    TSapling_transaction,
    // Artificial Types
    TRecord,
    TVariant,
};

export default Types;
