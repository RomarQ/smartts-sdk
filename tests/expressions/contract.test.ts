import {
    ContractStorage,
    Transfer,
    GetContract,
    Address,
    Mutez,
    Unit,
    ToContract,
    GetSome,
    GetVariable,
    ToAddress,
    GetSelf,
} from '../../src/expression';
import { Contract, EntryPoint } from '../../src/core';
import { verifyContractCompilationOutput } from '../util';
import { NewVariable, SetValue } from '../../src/statement';
import { TAddress, TUnit } from '../../src/type';

describe('Contract expressions', () => {
    it('GetContract', () => {
        const contract = new Contract().addEntrypoint(
            new EntryPoint('entry_point_1').code(() => [
                Transfer(
                    GetContract(Address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'), 'entry_point_1', TUnit()),
                    Mutez(100),
                    Unit(),
                ).send(),
            ]),
        );

        verifyContractCompilationOutput(contract);
    });
    it('ToContract', () => {
        const contract = new Contract().addEntrypoint(
            new EntryPoint('entry_point_1').code(() => [
                NewVariable(
                    'contract',
                    GetSome(ToContract(Address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'), 'entry_point_1', TUnit())),
                ),
                Transfer(GetVariable('contract'), Mutez(100), Unit()).send(),
            ]),
        );

        verifyContractCompilationOutput(contract);
    });
    it('ToAddress', () => {
        const contract = new Contract()
            .setStorageType(TAddress())
            .addEntrypoint(
                new EntryPoint('entry_point_1').code(() => [
                    SetValue(ContractStorage(), ToAddress(GetSelf('entry_point_1'))),
                ]),
            );

        verifyContractCompilationOutput(contract);
    });
});
