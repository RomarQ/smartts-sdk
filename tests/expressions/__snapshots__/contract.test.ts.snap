// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract expressions GetContract 1`] = `
"
        (
            template_id (static_id 0 (\\"contract.test.ts\\" 34))
            storage ()
            storage_type ((unknown 0))
            messages ((entry_point_1 True False False True (\\"contract.test.ts\\" 35) ((set_type (params (\\"contract.test.ts\\" 35)) \\"unit\\" (\\"contract.test.ts\\" 44)) (defineLocal \\"contract\\" (openVariant (contract entry_point_1 \\"unit\\" (literal (address KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT) (\\"contract.test.ts\\" 35)) (\\"contract.test.ts\\" 35)) \\"Some\\" None (\\"contract.test.ts\\" 35)) True (\\"contract.test.ts\\" 35)) (set (operations (\\"contract.test.ts\\" 35)) (cons (transfer (unit) (literal (mutez 100) (\\"contract.test.ts\\" 35)) (getLocal \\"contract\\" (\\"contract.test.ts\\" 35)) (\\"contract.test.ts\\" 35)) (operations (\\"contract.test.ts\\" 35)) (\\"contract.test.ts\\" 35)) (\\"contract.test.ts\\" 35)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"contract.test.ts\\" 34))
        )
        "
`;

exports[`Contract expressions GetContract 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %entry_point_1);\\\\nstorage   unit;\\\\ncode\\\\n  {\\\\n    CDR;        # @storage\\\\n    # == entry_point_1 ==\\\\n    # contract = sp.local(\\\\\\"contract\\\\\\", sp.contract(sp.TUnit, sp.address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'), entry_point='entry_point_1').open_some()) # @storage\\\\n    PUSH address \\\\\\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\\\\\"; # address : @storage\\\\n    CONTRACT %entry_point_1 unit; # option (contract unit) : @storage\\\\n    IF_NONE\\\\n      {\\\\n        PUSH int 35; # int : @storage\\\\n        FAILWITH;   # FAILED\\\\n      }\\\\n      {}; # @some : @storage\\\\n    # sp.transfer(sp.unit, sp.mutez(100), contract.value) # @some : @storage\\\\n    NIL operation; # list operation : @some : @storage\\\\n    SWAP;       # @some : list operation : @storage\\\\n    PUSH mutez 100; # mutez : @some : list operation : @storage\\\\n    UNIT;       # unit : mutez : @some : list operation : @storage\\\\n    TRANSFER_TOKENS; # operation : list operation : @storage\\\\n    CONS;       # list operation : @storage\\\\n    PAIR;       # pair (list operation) @storage\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%entry_point_1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CDR\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"address\\"
                            },
                            {
                                \\"string\\": \\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"CONTRACT\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"unit\\"
                            }
                        ],
                        \\"annots\\": [
                            \\"%entry_point_1\\"
                        ]
                    },
                    {
                        \\"prim\\": \\"IF_NONE\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"int\\"
                                        },
                                        {
                                            \\"int\\": \\"35\\"
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"FAILWITH\\"
                                }
                            ],
                            []
                        ]
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"SWAP\\"
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"mutez\\"
                            },
                            {
                                \\"int\\": \\"100\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"UNIT\\"
                    },
                    {
                        \\"prim\\": \\"TRANSFER_TOKENS\\"
                    },
                    {
                        \\"prim\\": \\"CONS\\"
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ]
}"
`;

exports[`Contract expressions ToAddress 1`] = `
"
        (
            template_id (static_id 0 (\\"contract.test.ts\\" 47))
            storage ()
            storage_type (\\"address\\")
            messages ((entry_point_1 True False False True (\\"contract.test.ts\\" 50) ((set_type (params (\\"contract.test.ts\\" 50)) \\"unit\\" (\\"contract.test.ts\\" 55)) (set (data) (to_address (self \\"entry_point_1\\" (\\"contract.test.ts\\" 50)) (\\"contract.test.ts\\" 50)) (\\"contract.test.ts\\" 50)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"contract.test.ts\\" 47))
        )
        "
`;

exports[`Contract expressions ToAddress 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %entry_point_1);\\\\nstorage   address;\\\\ncode\\\\n  {\\\\n    DROP;       # \\\\n    # == entry_point_1 ==\\\\n    # self.data = sp.self_entry_point_address('entry_point_1') # \\\\n    SELF %entry_point_1; # @self\\\\n    ADDRESS;    # address\\\\n    NIL operation; # list operation : address\\\\n    PAIR;       # pair (list operation) address\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"address\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%entry_point_1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"DROP\\"
                    },
                    {
                        \\"prim\\": \\"SELF\\",
                        \\"annots\\": [
                            \\"%entry_point_1\\"
                        ]
                    },
                    {
                        \\"prim\\": \\"ADDRESS\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ]
}"
`;

exports[`Contract expressions ToContract 1`] = `
"
        (
            template_id (static_id 0 (\\"contract.test.ts\\" 21))
            storage ()
            storage_type ((unknown 0))
            messages ((entry_point_1 True False False True (\\"contract.test.ts\\" 22) ((set_type (params (\\"contract.test.ts\\" 22)) \\"unit\\" (\\"contract.test.ts\\" 31)) (set (operations (\\"contract.test.ts\\" 22)) (cons (transfer (unit) (literal (mutez 100) (\\"contract.test.ts\\" 22)) (openVariant (contract entry_point_1 \\"unit\\" (literal (address KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT) (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)) \\"Some\\" (variant \\"Some\\" (literal (string \\"CONTRACT_NOT_FOUND\\") (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)) (operations (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)) (\\"contract.test.ts\\" 22)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"contract.test.ts\\" 21))
        )
        "
`;

exports[`Contract expressions ToContract 2`] = `
"{
    \\"micheline\\": \\"parameter (unit %entry_point_1);\\\\nstorage   unit;\\\\ncode\\\\n  {\\\\n    CDR;        # @storage\\\\n    # == entry_point_1 ==\\\\n    # sp.transfer(sp.unit, sp.mutez(100), sp.contract(sp.TUnit, sp.address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'), entry_point='entry_point_1').open_some(message = sp.some('CONTRACT_NOT_FOUND'))) # @storage\\\\n    NIL operation; # list operation : @storage\\\\n    PUSH address \\\\\\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\\\\\"; # address : list operation : @storage\\\\n    CONTRACT %entry_point_1 unit; # option (contract unit) : list operation : @storage\\\\n    IF_NONE\\\\n      {\\\\n        PUSH (option string) (Some \\\\\\"CONTRACT_NOT_FOUND\\\\\\"); # option string : list operation : @storage\\\\n        FAILWITH;   # FAILED\\\\n      }\\\\n      {}; # @some : list operation : @storage\\\\n    PUSH mutez 100; # mutez : @some : list operation : @storage\\\\n    UNIT;       # unit : mutez : @some : list operation : @storage\\\\n    TRANSFER_TOKENS; # operation : list operation : @storage\\\\n    CONS;       # list operation : @storage\\\\n    PAIR;       # pair (list operation) @storage\\\\n  };\\",
    \\"json\\": [
        {
            \\"prim\\": \\"storage\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\"
                }
            ]
        },
        {
            \\"prim\\": \\"parameter\\",
            \\"args\\": [
                {
                    \\"prim\\": \\"unit\\",
                    \\"annots\\": [
                        \\"%entry_point_1\\"
                    ]
                }
            ]
        },
        {
            \\"prim\\": \\"code\\",
            \\"args\\": [
                [
                    {
                        \\"prim\\": \\"CDR\\"
                    },
                    {
                        \\"prim\\": \\"NIL\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"operation\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"address\\"
                            },
                            {
                                \\"string\\": \\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"CONTRACT\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"unit\\"
                            }
                        ],
                        \\"annots\\": [
                            \\"%entry_point_1\\"
                        ]
                    },
                    {
                        \\"prim\\": \\"IF_NONE\\",
                        \\"args\\": [
                            [
                                {
                                    \\"prim\\": \\"PUSH\\",
                                    \\"args\\": [
                                        {
                                            \\"prim\\": \\"option\\",
                                            \\"args\\": [
                                                {
                                                    \\"prim\\": \\"string\\"
                                                }
                                            ]
                                        },
                                        {
                                            \\"prim\\": \\"Some\\",
                                            \\"args\\": [
                                                {
                                                    \\"string\\": \\"CONTRACT_NOT_FOUND\\"
                                                }
                                            ]
                                        }
                                    ]
                                },
                                {
                                    \\"prim\\": \\"FAILWITH\\"
                                }
                            ],
                            []
                        ]
                    },
                    {
                        \\"prim\\": \\"PUSH\\",
                        \\"args\\": [
                            {
                                \\"prim\\": \\"mutez\\"
                            },
                            {
                                \\"int\\": \\"100\\"
                            }
                        ]
                    },
                    {
                        \\"prim\\": \\"UNIT\\"
                    },
                    {
                        \\"prim\\": \\"TRANSFER_TOKENS\\"
                    },
                    {
                        \\"prim\\": \\"CONS\\"
                    },
                    {
                        \\"prim\\": \\"PAIR\\"
                    }
                ]
            ]
        }
    ]
}"
`;
